### DISCUSSION TITLE
Using a macro to reduce boilerplate for registering a dependency

### DISCUSSION THREAD

**Discussion by Jeehut** [pointfreeco/swift-dependencies#116] [Ideas]
I just registered a custom dependency by writing the following boilerplate code in my existing `APIClient.swift` file:

```swift
struct APIClient { ... }

private enum APIClientKey: DependencyKey {
   static let liveValue = APIClient()
}

extension DependencyValues {
  var apiClient: APIClient {
    get { self[APIClientKey.self] }
    set { self[APIClientKey.self] = newValue }
  }
}
```

While this is kinda manageable, I tend to define multiple dependencies per project and always need to look up and copy & paste the syntax on how to define these as I can't really memorize the type names `DependencyKey` and `DependencyValues`. And I have a feeling that the above boilerplate code (especially the extension) could be generated by a macro. I would imagine something like this:

```swift
@Dependency(name: "apiClient", values: [.live])
struct APIClient { ... }
```

I'm not sure if extending types or creating new types is supported by macros (yet), but any improvements that make the registration of custom dependencies easier and less type- and lookup-intensive would be appreciated!

The suggested syntax with the `values` could be replaced by a convention-based mechanism, automatically detecting `live`, `test`, or other static properties of type `Self` within the dependency type. When no `values` are specified, only `live` could be automatically detected if the type has an initializer without parameters. Just a few first thoughts to get started.

**Comment by mbrandonw** [comment]
Hi @Jeehut, I'm not 100% certain, but I believe the macro syntax you have theorized can't actually work with macros today. 

Another wrinkle to take into account is that adding swift-syntax to a package incurs a 30 second compilation penalty, which is pretty significant for a library that currently compiles in less than 2 seconds. That means it would probably have to vended in a separate "DependenciesMacros" library, or the macro would have to be written without using swift-syntax, which can be difficult.

We're certainly open to reducing the boilerplate for registering a dependency, but at the end of the day it is also quite insignificant compared to the code that is written everyday. If you end up experimenting with what's possible please do share in this discussion.

**Reply by Jeehut** [reply]
Thank you for explaining your point of view. I will make sure to share any discoveries.

**Reply by mbrandonw** [reply]
Here's a macro that I think could actually be implemented with Swift macros today:

```swift
@DependencyKey("apiClient")
extension DependencyValues {
  static var liveValue: APIClient { … }
  static var previewValue: APIClient { … }
}

// expands to

extension DependencyValues {
  var apiClient: APIClient { get { … } set { … } }
  private enum APIClient_Key_uniqueID: DependencyKey {
    static var liveValue: APIClient { … }
    static var previewValue: APIClient { … }
  }
}
```

What this does is allow you to attach the macro to an extension of `DependencyValues` (and you need to provide the name of the property you want to add). Then in that extension you specify the live/preview/test values. When the macro expands the code, it creates a new private type inside the extension and moves the live/preview/test values to that type. And it also adds the computed property.

And you would probably also need a `@TestDependencyKey` macro.

This would work OK, but honestly I find it weird that we are defining `liveValue`/`previewValue`/`testValue` inside `DependencyValues`, but they don't actually stay there. They get moved to some separate type. That is pretty surprising, and requires the user to know how the macro works in order to understand by `DependencyValues.liveValue` is not valid.

Anyway, just wanted to share one possible direction for a macro as imperfect as it is.

**Reply by Jeehut** [reply]
Thank you for sharing your exploration. Unfortunately, if the macro is applied to `DependencyValues` and is called `DependencyKey`, I still need to remember those names. If there was a dictionary involved in the API, I probably would easily remember those, but for a `Dependency` it's kinda hard, for me at least. My suggestion has the advantage of only sticking a `Dependency` macro (which is pretty easy to remember) to the dependency type directly. But that doesn't seem to work, so I'll need to further think about this...

I'm currently exploring macros for my error-handling library, so maybe I'll have some ideas once that is complete. 🤞

**Reply by mbrandonw** [reply]
I don’t think we can ever make it so that people don’t think about `DependencyKey` since one often needs to conform to it separately from `TestDependencyKey`. But `DependencyValues` is a type that one does not often need to deal with in practice.

**Reply by Jeehut** [reply]
Getting rid of `DependencyValues` already would be an improvement, I think.

**Comment by benrudhart** [comment]
I'm also wondering why there is no macro for this one. Since TCA by now already comes with a bunch of macros (like @Reducer - Thank you 👏) I  would've expected a macro for the DependencyKeys, similar to the @Entry one from SwiftUI.
Or did I miss something?

**Reply by stephencelis** [reply]
@benrudhart There's more discussion specific to `@Entry` here: https://github.com/pointfreeco/swift-dependencies/discussions/258

The short of it is we're open to proposals for a similar macro, but we don't see how such a macro is possible or useful in the long term given the extra requirement (`testValue` _and_ `liveValue`).

If you (or someone else) comes up with something interesting and useful, though, we'd love to see it!

**Reply by benrudhart** [reply]
I see, will check that one. Thank you.
